docarray/array/storage/milvus/backend.py

```diff
@@ -0,0 +1,320 @@
+            *extra_columns,
+        ]
+
+    @staticmethod
+    def _docs_from_query_response(response):
+        return DocumentArray([Document.from_base64(d['serialized']) for d in response])
+
+    @staticmethod
+    def _docs_from_search_response(
+        responses,
+    ) -> 'Union[List[DocumentArray], DocumentArray]':
+        das = []
+        for r in responses:
+            das.append(
+                DocumentArray(
+                    [Document.from_base64(hit.entity.get('serialized')) for hit in r]
+                )
+            )
+        return das if len(das) > 0 else das[0]
+
+    def _update_consistency_level(self, **kwargs):
+        kwargs_consistency_level = kwargs.get('consistency_level', None)
+        kwargs['consistency_level'] = (
+            kwargs_consistency_level
+            if kwargs_consistency_level
+            else self._config.consistency_level
+        )
+        return kwargs
+
+    def _map_embedding(self, embedding):
+        if embedding is not None:
+            from docarray.math.ndarray import to_numpy_array
+
+            embedding = to_numpy_array(embedding)
+
+            if embedding.ndim > 1:
+                embedding = np.asarray(embedding).squeeze()
+        else:
+            embedding = np.zeros(self._config.n_dim)
+        return embedding
+
+    def __getstate__(self):
+        d = dict(self.__dict__)
+        del d['_collection']
+        del d['_offset2id_collection']
+        return d
+
+    def __setstate__(self, state):
+        self.__dict__ = state
+        self._collection = self._create_or_reuse_collection()
+        self._offset2id_collection = self._create_or_reuse_offset2id_collection()
+
+    def __enter__(self):
+        _ = super().__enter__()
+        self._collection.load()
+        self._offset2id_collection.load()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._collection.release()
+        self._offset2id_collection.release()
+        super().__exit__(exc_type, exc_val, exc_tb)
+
+    def loaded_collection(self, collection=None):
+        """
+        Context manager to load a collection and release it after the context is exited.
+        If the collection is already loaded when entering, it will not be released while exiting.
+
+        :param collection: the collection to load. If None, the collection of this indexer is used.
+        :return: Context manager for the provided collection.
+        """
+
+        class LoadedCollectionMngr:
+            def __init__(self, coll, connection_alias):
+                self._collection = coll
+                self._loaded_when_enter = False
+                self._connection_alias = connection_alias
+
+            def __enter__(self):
+                self._loaded_when_enter = (
+                    loading_progress(
+                        self._collection.name, using=self._connection_alias
+                    )['loading_progress']
+                    != '0%'
```

Comment:
this does not look good at all

Reply:
Yeah, I hate it. Unfortunately it seems to be the only way that the Milvus API exposes.

Reply:
Gigachad ask us to shame this :P

Reply:
Can u explain what it tries to do?

Reply:
So this context manager loads the collection into memory and releases it at the end, and it's used in `find()` and a bunch of other methods.

The `with da:` context manager also performs this function, so that users can bunch together multiple operations while loading/releasing only once.

For that to work, this `loaded_collection` manager has to detect during `__enter__` if the collection is already loaded. In this case it should not be released at exit, because the `with da` manager takes care of that.

Unfortunately, `collection.is_loaded()` is not a thing that exists in pymilvus, and this seems to be the only way to detect if a collection is loaded or not.
The reasoning behind the `!= '0%'` check (as opposed to `== '100%'`) is that any status between 0 and 100 means that some entity higher up in the stack is messing with loading behaviour, and thus we should let it handle releasing as well.

Reply:
We wait for the loading to be completed, so no error should arise from this.